File: backend/app/api/routes/company.py
"""
Company Statistics API Routes

Provides company profile, subsidiaries, leadership, shareholders, and capital/dividend data
from SSI iBoard API.
"""

from typing import List, Optional
from fastapi import APIRouter, HTTPException
import logging

from app.services.iboard_client import get_iboard_client

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/company", tags=["Company"])


@router.get("/{symbol}/profile")
async def get_company_profile(symbol: str):
    """Get company profile information.
    
    Returns company info including:
    - Basic info: name, industry, founding date
    - Financial: charter capital, listed value
    - Contact: address, website, email, phone
    - Trading: exchange, listing date, stock type
    
    Cache: 1 hour
    """
    client = get_iboard_client()
    data = await client.get_company_profile(symbol)
    
    if data is None:
        raise HTTPException(status_code=404, detail=f"Company profile not found for {symbol}")
    
    return data


@router.get("/{symbol}/subsidiaries")
async def get_subsidiaries(symbol: str):
    """Get list of subsidiaries and affiliated companies.
    
    Returns list of companies with:
    - childSymbol, childCompanyName
    - percentage ownership
    - roleName (e.g., "Công ty con", "Công ty liên kết")
    
    Cache: 1 hour
    """
    client = get_iboard_client()
    data = await client.get_sub_companies(symbol)
    
    if data is None:
        return []  # Empty list if no subsidiaries
    
    return data


@router.get("/{symbol}/leadership")
async def get_leadership(symbol: str):
    """Get list of company leadership (board members, executives).
    
    Returns list of leaders with:
    - fullName
    - positionName (e.g., "Chủ tịch HĐQT", "Tổng Giám đốc")
    
    Cache: 1 hour
    """
    client = get_iboard_client()
    data = await client.get_company_leadership(symbol)
    
    if data is None:
        return []
    
    return data


@router.get("/{symbol}/shareholders")
async def get_shareholders(symbol: str):
    """Get detailed list of shareholders (individual and institutional).
    
    Returns list of shareholders with:
    - name, quantity, percentage
    - type: "I" (Individual) or "C" (Corporation)
    - ownerShipTypeCode: "NGOAI" (Foreign), "CNNHAN" (Individual), "KHAC" (Other)
    
    Cache: 30 minutes
    """
    client = get_iboard_client()
    data = await client.get_shareholder_detail(symbol)
    
    if data is None:
        return []
    
    return data


@router.get("/{symbol}/shareholder-summary")
async def get_shareholder_summary(symbol: str):
    """Get shareholder structure summary (for pie chart).
    
    Returns summary with:
    - foreignerPercentage
    - statePercentage
    - otherPercentage
    
    Cache: 30 minutes
    """
    client = get_iboard_client()
    data = await client.get_shareholder_summary(symbol)
    
    if data is None:
        raise HTTPException(status_code=404, detail=f"Shareholder summary not found for {symbol}")
    
    return data


@router.get("/{symbol}/cap-dividend")
async def get_cap_dividend(symbol: str):
    """Get capital and dividend history.
    
    Returns:
    - assetList: Yearly total assets
    - cashDividendList: Yearly cash dividend per share
    
    Cache: 1 hour
    """
    client = get_iboard_client()
    data = await client.get_cap_and_dividend(symbol)
    
    if data is None:
        return {"assetList": [], "cashDividendList": []}
    
    return data

File: frontend/src/types/company.ts
/**
 * Company Data Types - SSI iBoard API
 * 
 * Types for company profile, subsidiaries, leadership, shareholders, and capital/dividend data
 */

// ============================================
// Company Profile
// ============================================

export interface CompanyProfile {
    symbol: string;
    companyName: string;
    industryName: string;
    superSector: string;
    sector: string;
    subSector: string;
    foundingDate: string;
    charterCapital: string;
    numberOfEmployee: number;
    companyProfile: string; // HTML description
    listingDate: string;
    exchange: string;
    firstPrice: string;
    issueShare: string;
    listedValue: string;
    quantity: number;
    stockType: string;
    freeFloatRate: string;
    updateDate: string;
    address: string;
    website: string;
    telephone: string;
    email: string;
    fax: string | null;
}

// ============================================
// Subsidiaries
// ============================================

export interface SubCompany {
    parentSymbol: string;
    parentCompanyName: string;
    childSymbol: string;
    childCompanyName: string;
    charterCapital: string | null;
    percentage: string | null;
    roleId: string;
    roleName: string;
}

// ============================================
// Leadership
// ============================================

export interface LeadershipMember {
    symbol: string;
    fullName: string;
    positionName: string;
    positionLevel: string;
    positionId: string;
    personId: string;
}

// ============================================
// Shareholders
// ============================================

/** Individual shareholder details */
export interface ShareholderDetail {
    symbol: string;
    name: string;
    quantity: number;
    percentage: number;
    publicDate: string;
    ownerShipTypeCode: "NGOAI" | "CNNHAN" | "KHAC" | string;
    type: "I" | "C"; // Individual / Corporation
    totalPage: number;
    totalRow: number;
}

/** Summary of shareholder structure (for pie chart) */
export interface ShareholderSummary {
    symbol: string;
    foreignerVolume: string;
    foreignerPercentage: string;
    stateVolume: string;
    statePercentage: string;
    otherVolume: string;
    otherPercentage: string;
    publicDate: string;
}

// ============================================
// Capital & Dividend
// ============================================

export interface AssetYearly {
    year: string;
    asset: string;
}

export interface CashDividendYearly {
    year: string;
    valuePershare: string;
}

export interface CapAndDividend {
    assetList: AssetYearly[];
    cashDividendList: CashDividendYearly[];
}

// ============================================
// API Response Wrappers
// ============================================

export interface CompanyApiResponse<T> {
    code: string;
    message: string;
    data: T;
    paging: { page: number; pageSize: number; totalPage: number; totalRow: number } | null;
    status: string;
}
